### PR TITLE
Transposition Cipher Lab(s)

### PR DESCRIPTION
**Pull request by WeeknightMVP**
It took me a day despite @weaversa spotting me the blueprint for partitioning sequences over a predicate, but I have finally managed to specify the humble [Rail Fence](https://en.wikipedia.org/wiki/Transposition_cipher#Rail_Fence_cipher).  Hopefully [Route](https://en.wikipedia.org/wiki/Transposition_cipher#Route_cipher) and other transposition ciphers will be easier...

### CODE REVIEW

File: labs/Demos/RailFence.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+/** Swap `i`th and `j`th entries of sequence `a` */
+swap : {n, i, a} (fin i, i == width n + 1) => [n]a -> [i] -> [i] -> [n]a
+swap a i j = updates a [i,j] (a @@ [j,i])
+
+/** indices per cycle */
+cycle: {r} _r_ r => ([cw r], [r-1][2][cw r], [cw r])
+cycle = (0, transpose [take`{r-1} ([1...]:[inf][cw r]), reverse (take`{r-1} ([`(r+1)...]:[inf][cw r]))], `r)
+
+/** inverse a 1:1 mapping of permuted values [0..`(n-1)] */
```
Comment (weaversa) [inline labs/Demos/RailFence.md:22]: “inverse a”?

Comment (WeeknightMVP) [reply labs/Demos/RailFence.md:22]: Nouning verbs is all the rage. I still need to write a narrative around this.

**Comment by WeeknightMVP** [comment]
I have pushed substantial changes to the lab(s) on transposition ciphers.  They are meant to be visited in the following order:
* [CommonProperties](/weaversa/cryptol-course/tree/lab-transposition/labs/Transposition/CommonProperties.md) (an offshoot of `CryptoProofs` treating higher-order functions and properties)
* [Transposition](/weaversa/cryptol-course/tree/lab-transposition/labs/Transposition/Transposition.md) (a library defining common concepts for transposition ciphers; most notably permutation mappings and supporting functions and properties)
* [Esrever](/weaversa/cryptol-course/tree/lab-transposition/labs/Transposition/Esrever.md) (a trivial reverse "cipher" demonstrating how to specify permutation mappings for the encrypt and decrypt functions from Transposition)
* [Scytale](/weaversa/cryptol-course/tree/lab-transposition/labs/Transposition/Scytale.md) (an adaptation of the Scytale cipher described in Programming Cryptol)
* [RailFence](/weaversa/cryptol-course/tree/lab-transposition/labs/Transposition/RailFence.md) (a specification of the Rail Fence cipher, which is easy on paper but hard in Cryptol; still in progress)
* [Route](/weaversa/cryptol-course/tree/lab-transposition/labs/Transposition/Route.md) (an unguided suggestion for sufficiently motivated course participants to define the Route cipher using concepts learned in prior labs; not yet included)

RailFence, Route, and prerequisites for each lab still need to be filled in.

**Comment by WeeknightMVP** [comment]
Almost there.  Specifying that this `Scytale` spec agrees with Galois' was a challenge.  `RailFence` is done.  Just need to add some boilerplate for `Route` and regenerate the non-Answers files.  There's enough here to be ready for review.

**Comment by WeeknightMVP** [comment]
Yeah, the Travis CI scripts are failing because the Cryptol specs use Cryptol 2.9 typeclasses.  The short-term fix is to change these typeclasses to Cryptol 2.8; the long-term fix is to revert these changes and update the rest of the course to 2.9...

**Comment by weaversa** [comment]
Let's plan to move to 2.9 after next week is over. I don't think we're
gonna merge this branch before next week, so you can leave it as it is. In
the meantime, we could update TravisCI to use 2.9 :-)

**Comment by WeeknightMVP** [comment]
A lot of it was too generic anyway.  It should be easy to go back to 2.9, since I changed `w` to `ix` to reflect 2.8 idioms.

**Comment by WeeknightMVP** [comment]
Wow, some of these proofs in `TranspositionAnswers.md` are dragging in 2.9.  ``:prove permute_injective`{32, Char, [width 16]}`` was taking about 10s, but now ``:prove permute_injective`{16, Char, Integer}`` is taking...longer than I am willing to wait before writing this comment...

**Review by weaversa — APPROVED** [review]
I'm happy to merge this in even without the prereqs being totally filled in. Great work!

**Pull request merged**